### ISSUE TITLE
S2259 False Positive using extension method

### ISSUE THREAD

**Issue by evanbb** [SonarSource/sonar-dotnet#3850]
### Description

Appears to be related to #2639 

When creating a custom null validation method with a `[ValidatedNotNullAttribute]` marking the parameter in question, there appears to be a discrepancy based on how the method is called. If invoked directly, Sonar correctly recognizes that `value` cannot be null. However, a false positive is returned when the null-checking method is invoked as an extension method:

### Repro steps

```cs
namespace SonarS2259FalsePositive
{
    public static class ListExtensions
    {
        public static bool HasItems<T>([ValidatedNotNull] this List<T> list)
        {
            return !(list is null) && list.Count > 0;
        }
    }

    public static class ObjectExtensions
    {
        public static void AssertNotNull([ValidatedNotNull] this object obj)
        {
            if (obj is null)
            {
                throw new ArgumentNullException(nameof(obj));
            }
        }
    }

    public class SonarFalsePositive
    {
        public IEnumerable<int> HasItemsExtension(Wrapper wrapper)
        {
            List<int> items = wrapper.List?.ListOfInts;

            if (items.HasItems())
            {
                foreach (var t in items) // <-- this gives a false positive
                {
                    yield return t;
                }
            }
        }

        public IEnumerable<int> HasItemsDirect(Wrapper wrapper)
        {
            List<int> items = wrapper.List?.ListOfInts;

            if (ListExtensions.HasItems(items))
            {
                foreach (var t in items) // <-- this works as expected
                {
                    yield return t;
                }
            }
        }

        public IEnumerable<int> NotNullExtension(Wrapper wrapper)
        {
            List<int> items = wrapper.List?.ListOfInts;

            items.AssertNotNull();

            foreach (var t in items) // <-- this gives a false positive
            {
                yield return t;
            }
        }

        public IEnumerable<int> NotNullDirect(Wrapper wrapper)
        {
            List<int> items = wrapper.List?.ListOfInts;

            ObjectExtensions.AssertNotNull(items);

            foreach (var t in items) // <-- this works as expected
            {
                yield return t;
            }
        }
    }

    public class SomethingWithAListOfInts
    {
        public List<int> ListOfInts { get; set; }
    }

    public class Wrapper
    {
        public SomethingWithAListOfInts List { get; set; }
    }
}
```

### Expected behavior

Rule should not be violated when null-checking method is invoked as an extension method.

### Actual behavior

Rule is violated when null-checking method is invoked as an extension method.

### Known workarounds

Invoke null-checking method directly

### Related information

* SonarQube version
  * 7.9.4 (build 35981)
* C# Plugin Version
  * 8.6.1

**Comment by pavel-mikula-sonarsource** [comment]
Hi @evanbb ,

Thank you for reporting this case. I can confirm it as False Positive. The root cause is the same as in #2639 as you've mentioned.
